lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableCache.java

```diff
@@ -0,0 +1,194 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base.cache;
+
+import com.google.common.cache.AbstractCache;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheStats;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.CheckForNull;
+
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static io.airlift.concurrent.MoreFutures.getDone;
+import static java.lang.System.nanoTime;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A Cache implementation similar to ones produced by {@code com.google.common.cache.CacheBuilder},
+ * but one that does not exhibits <a href="https://github.com/google/guava/issues/1881">Guava issue #1881</a>, i.e.
+ * a {@link #getIfPresent(Object)} after {@link #invalidate(Object)} is guaranteed to return {@code null} and
+ * {@link #get(Object, Callable)} after {@link #invalidate(Object)} is guaranteed to load a fresh value.
+ */
+public class EvictableCache<K, V>
+        extends AbstractCache<K, V>
+        implements Cache<K, V>
+{
+    /**
+     * @apiNote Piggy-back on {@link CacheBuilder} for cache TTL.
+     */
+    public static <K, V> EvictableCache<K, V> buildWith(CacheBuilder<? super K, Object> cacheBuilder)
+    {
+        return new EvictableCache<>(cacheBuilder);
+    }
+
+    // private final Map<K, Future<V>> map = new ConcurrentHashMap<>();
+    private final Cache<K, Future<V>> delegate;
+
+    private final StatsCounter statsCounter = new SimpleStatsCounter();
+
+    public EvictableCache(CacheBuilder<? super K, Object> cacheBuilder)
+    {
+        this.delegate = cacheBuilder.build();
+    }
+
+    @CheckForNull
+    @Override
+    public V getIfPresent(Object key)
+    {
+        Future<V> future = delegate.getIfPresent(key);
+        if (future != null && future.isDone()) {
+            statsCounter.recordHits(1);
+            return getDone(future);
+        }
+        return null;
+    }
+
+    @Override
+    public V get(K key, Callable<? extends V> loader)
+            throws ExecutionException
+    {
+        requireNonNull(key, "key is null");
+        requireNonNull(loader, "loader is null");
+
+        while (true) {
+            SettableFuture<V> newFuture = SettableFuture.create();
```

Comment:
Should we use `CompletableFuture` in new code?

Reply:
I don't think we should. We stick with Guava's family of futures due to how cancellations are propagated. It doesn't matter here, but your question is a generic one.